### ISSUE TITLE
Race between InitTmpDir and GetTempFileDirectory in TestDistributed

### ISSUE THREAD

**Issue by dummyunit** [fastbuild/fastbuild#600]
Event sequence that leads to the race:
1. `TestDistributed::TestHelper` creates `Server`, which creates `JobQueueRemote`, which starts worker threads.
2. `TestDistributed::TestHelper` then calls `FBuild::Build`, which creates `JobQueue`.
3. `JobQueue` calls `WorkerThread::InitTmpDir`, which writes to `WorkerThread::s_TmpRoot`, this happens on the main thread.
4. Worker threads at the start call `CreateThreadLocalTmpDir`, which calls `GetTempFileDirectory`, which reads from `WorkerThread::s_TmpRoot`.

This was found by ThreadSanitizer, full log is [here](https://gist.github.com/dummyunit/07a56143619e89709201e11f9913bbe6) (line numbers are incorrect, don't know why).

As far as I understand this problem doesn't affect `FBuild` and `FBuildWorker` because they create ether `JobQueue` or `JobQueueRemote` but not both, and so `InitTmpDir` is called only once before `s_TmpRoot` is used.

It seems that protecting access to `s_TmpRoot` with a mutex would fix the race in tests, but I don't know if that would be acceptable because it may affect performance of `FBuild` and `FBuildWorker` where the race doesn't happen.

**Comment by ffulin** [comment]
Great find!

I've implemented your suggested fix (https://github.com/fastbuild/fastbuild/commit/90657f697f9ff2457373e24cc0f57d366f3e308e). The variable is infrequently accessed and this lock will be rarely contested, so there should be no performance impact.

(I looked briefly into extracting the variable so it's no longer shared between the queues, but it's a little entangled as it's used by code that is common to both)
